matrix/matrix.py

```
@@ -0,0 +1,198 @@
+import operator
+from itertools import repeat
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type. This
+    implementation only supports even-side matrices (n * n).
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+                         the matrix is represented by a tuple.
+    """
+    __slots__ = '_matrix'
+
+    def __init__(self, *lines):
+        if not self._is_valid_data(*lines):
```

"data is not valid" is not a good error message.

Instead of checking the return value of `_is_valid_data`, simply call it `validate` and let it throw the `ValueError` with the correct message.

Less code, more descriptive errors